### ISSUE TITLE
Go Module

### ISSUE THREAD

**Issue by cimnine** [go-resty/resty#178]
The documentation suggests that one should use resty as follows:

```
import "gopkg.in/resty.v1"
```

(https://github.com/go-resty/resty/blob/ec17de1c59a99e0743dac77aaab6435f364fdc42/README.md#usage)

But this brakes with Go modules, because `go.mod` declares the module as follows:

https://github.com/go-resty/resty/blob/ec17de1c59a99e0743dac77aaab6435f364fdc42/go.mod#L1

This causes troubles when I try to compile code with go11.1rc2 and go modules enabled.

All is fine when I import resty like this:

```go
import "github.com/go-resty/resty"
```

Now there are two options:

a) Adjust the documentation and suggest to `import "github.com/go-resty/resty"`
b) Adjust the `go.mod` to be `module "gopkg.in/resty.v1`

Option (a) is good for all new project that start with go modules right away, but it's backward incompatible.
Option (b) would be good for projects that used plain `go get` or `dep` until now and upgrade to `go.mod`, because they don't have to change all the import statements. There are no further drawbacks, except that this option needs a new release with an updated `go.mod`.

**Comment by jeevatkm** [comment]
@cimnine I would like to clarify if understood correctly.

You mean to say using `gopkg.in/resty.v1` with Go modules breaks the build, correct?

----

By nature of resty package it will work seamlessly with `github.com/go-resty/resty` and `gopkg.in/resty.v1` in plain `go get` (on go1.11 it has to have `GO111MODULE=off`). It keeps backward compatibility with plain go get. Could you please try and let me know?

But if we use `gopkg.in/resty.v1` with Go modules it will break, as you have mentioned `go.mod` defines it as `github.com/go-resty/resty`. This is expected behavior.

FYI, resty package currently does not enforce package import like `package resty // import "github.com/go-resty/resty"`. I'm planning to do that for `v2.0.0` release in favor of Go modules.

I understand, documentation on `Usage` section needs update for Go modules instructions. I will do (or you can also send PR).  Thank you for bringing it up.

**Comment by cimnine** [comment]
> You mean to say using gopkg.in/resty.v1 with Go modules breaks the build, correct?

Yes.

> github.com/go-resty/resty

This will work with plain `go get` as well.

> I understand, documentation on Usage section needs update for Go modules instructions.

Yes.

But: Everybody who updates their current project to Go modules will have to adjust all import statements in the project from `gopkg.in/resty.v1` to `github.com/go-resty/resty`. This would not be necessary, if `go.mod` would declare the project as `gopkg.in/resty.v1`.

---

What you do in v2 is up to you. The Go community suggests for major version increments to adjust the import path anyway, so switching from `gopkg.in/...` to `github.com/...` would not be extra effort.

**Comment by jeevatkm** [comment]
> Everybody who updates their current project to Go modules will have to adjust all import statements in the project from gopkg.in/resty.v1 to github.com/go-resty/resty. This would not be necessary, if go.mod would declare the project as gopkg.in/resty.v1.

Okay, let's update the `go.mod` to `gopkg.in/resty.v1` and documentation update 😄

**Comment by cimnine** [comment]
It's hard to test, because you'd have to release a new release to gopkg.in with the new `go.mod`. It's afaik not possible to clone reference master through gopkg.in.

**Comment by jeevatkm** [comment]
@cimnine Yes you're correct. Directly we cannot get master via gopkg.in. But we can workaround that.

I believe you have `gopkg.in/resty.v1` on your GOPATH. So goto `$GOPATH/src/gopkg.in/resty.v1` and do `git fetch && git checkout master && git pull`. 

I'm at work, I will make a release by EOD when I get home. Thanks.

**Comment by jeevatkm** [comment]
Thanks. I will make a release by EOD.

**Comment by jeevatkm** [comment]
Making `v1.9.0` release.

**Comment by vvarp** [comment]
When running `go mod tidy` on a project that uses resty as dependency, I am getting:

```
go: github.com/go-resty/resty@v1.9.0: parsing go.mod: unexpected module path "gopkg.in/resty.v1"
```

Not sure how to interpret that?

**Comment by jeevatkm** [comment]
@vvarp As part this #178 issue, go.mod is updated to `gopkg.in/resty.v1` so that existing resty users could migrate to Go modules without updating their import path to `github.com/go-resty/resty`. Please see the thread conversation and use `gopkg.in/resty.v1`.

On upcoming `v2.0.0` release I will default and enforce import path to Go modules v2 path.  I'm planning to remove distribution via gopkg.in v2 onward.

**Comment by vvarp** [comment]
Sorry, I forgot to mention that my go.mod is using `gopkg.in/resty.v1 v1.9.0`, however I still get the above when running `go mod tidy`

**Comment by jeevatkm** [comment]
@vvarp Is Go module build failing too? Also which import path are you using in your project (`gopkg.in/resty.v1` or `github.com/go-resty/resty`)?

@cimnine Could you please share your inputs from your testing?

**Comment by zuf** [comment]
I also have this issue.

in go.mod: 
```
require gopkg.in/resty.v1 v1.9.0
```

import in go code: 
```go
import "gopkg.in/resty.v1"
```

go version go1.11 linux/amd64
`go mod tidy` says 
```
go: github.com/go-resty/resty@v1.9.0: parsing go.mod: unexpected module path "gopkg.in/resty.v1"
go: error loading module requirements
```

**Comment by jeevatkm** [comment]
@zuf @vvarp I will investigate it and get back to you. If any changes required I will do it.

@cimnine Can you try `go mod tidy` on your end please?

**Comment by cimnine** [comment]
```
$ go mod tidy

$ 
```

It's all good.

**Comment by cimnine** [comment]
I don't think that's the cause, but the example test uses the wrong import path:

https://github.com/go-resty/resty/blob/master/example_test.go#L19

**Comment by jeevatkm** [comment]
@cimnine Thanks for your inputs. we both kind of doubted the same.

@vvarp @zuf I'm able to replicate the issue you have mentioned with `v1.9.0`.

What I did?

* I have cleaned `github.com/go-resty/resty` and `gopkg.in/resty.v1` from my GOPTAH. To make sure I don't have copies of resty.
* Created a simple main.go with resty get request and go.mod with `require gopkg.in/resty.v1 v1.9.0`
* All go mod commands success except `go mod tidy` and `go mod why`
* After the `go build`; I have verified the path `$GOPATH/pkg/mod/`; it had both version of resty (gopkg.in and github.com). 
* Then I did grep of `github.com/go-resty/resty` on source code found `example_test.go` using github.com import path. (This moment I got a notification from github of @cimnine comment)
* I have modified the `$GOPATH/pkg/mod/gopkg.in/resty.v1@v1.9.0/example_test.go`  with gopkg.in import path
* Now all `go mod` commands success and only gopkg.in version is exists in `$GOPATH/pkg/mod/`

> I did not expected `*_test.go` files would cause an issue with go mod. Typically test deps should be separate. 

Long story short, I have to update `example_test.go` with gopkg.in  import path and make a release of `v1.9.1`

I will make an release when I get back to home.

**Comment by jeevatkm** [comment]
@leonsim both of us understanding about go mod spec correct. And you're point is valid about not to use `v1` in the import path and I agree with you.

However as you might have read this thread. I'm not planning introduce breaking change in minor version (I did before and reverted in `v1.9.1`) of changing import path.  I'm plan to have breaking change in major version and moving away from `gopkg.in` in-favor of  go mod.

`go mod` implementation has support for `gopkg.in` import path(s), it should work flawlessly with `.vX` in the import path (even though its away from spec).  

Let's aim for `resty v2`. Thank you.

**Comment by leonsim** [comment]
Thank you, @jeevatkm 

After I update `go.mod` to `require gopkg.in/resty.v1 v1.9.1` and then `go mod tidy` works without any error.

**Comment by lanshipeng** [comment]
i have a similar problem,when i use go mod tidy or go build . it's not succeed.
![image](https://user-images.githubusercontent.com/21244175/49415283-2eb9ca00-f7b1-11e8-9280-411c175916bd.png)

**Comment by jeevatkm** [comment]
@lanshipeng Could you please update the import path to `gopkg.in/resty.v1` and add entry in the go.mod as `require gopkg.in/resty.v1 v1.9.1`

Then you should be good.

**Comment by lanshipeng** [comment]
@jeevatkm but i don't know where to update the import path ,i can't find where 
transfer the "gopkg.in/resty.v1 "

**Comment by jeevatkm** [comment]
@lanshipeng In your Go project search for `github.com/go-resty/resty` and replace it with `gopkg.in/resty.v1`

And if you have a file `go.mod` then open that file search for `github.com/go-resty/resty` and replace that line with `gopkg.in/resty.v1 v1.10.2`

**Comment by lanshipeng** [comment]
@jeevatkm ,thank you. after i replace "github.com/go-resty/resty" with  "gopkg.in/resty.v1 v1.10.2" ,it's run ok

**Comment by jeevatkm** [comment]
@lanshipeng You're welcome.

**Comment by lanshipeng** [comment]
@jeevatkm but when i use go build ./... it occured like this:
`go: gopkg.in/resty.v1@v1.10.2 used for two different module paths (github.com/go-resty/resty and gopkg.in/resty.v1)
`

**Comment by tsouza** [comment]
I can't seem to get this working. I have cleared everything related to go-resty under GOPATH also fixed the import to be `gopkg.in/resty.v1` and I got the following in `go.mod`: `gopkg.in/resty.v1 v1.12.0`. When I run `go mod tidy` I get:
```
go: github.com/go-resty/resty@v1.12.0: parsing go.mod: unexpected module path "gopkg.in/resty.v1"
```

**Comment by jeevatkm** [comment]
@tsouza Please have a look this one https://github.com/go-resty/resty/issues/230#issuecomment-467911306 and try it out?

**Comment by tsouza** [comment]
@jeevatkm I have that fixed but still with the same issue. For now, I've dropped `go-resty` and I am using golang http.
